spec/index.bs

```diff
@@ -15742,17 +15762,78 @@ This stage produces a <dfn dfn>Fragment</dfn> for each [=RasterizationPoint=]:
+                    |depthStencilDesc|?.{{GPUDepthStencilState/depthCompare}}).
         1. If `sample_mask` [=builtin=] is produced by the shader as |value|:
             1. Set |fragment|.[=Fragment/coverageMask=] to |fragment|.[=Fragment/coverageMask=] &and; |value|.
 
         Otherwise we are in [[#no-color-output]] mode, and |fragment|.[=Fragment/colors=] is empty.
     1. Return |fragment|.
 </div>
 
+<div algorithm data-timeline=queue>
+    <dfn abstract-op>compare fragment</dfn>(destination, value, aspect, attachment, compareFunc)
+
+    **Arguments:**
+
+    - |destination|: The [=FragmentDestination=].
+    - |value|: The value to be compared.
+    - |aspect|: The aspect of |attachement| to sample values from.
+    - |attachment|: The attachment to be compared against.
+    - |compareFunc|: The comparison function to use.
+
+    **Returns:** `true` if the comparison passes, or `false` otherwise
+
+    - If |attachement| is `undefined` or does not have |aspect|, return `true`.
+    - If |compareFunc| is `undefined` or {{GPUCompareFunction/"always"}}, return `true`.
+    - Let |attachmentValue| be the value of |aspect| of |attachment| at |destination|.
+    - Return `true` if comparing |value| with |attachmentValue| using |compareFunc| succeeds, and `false` otherwise.
+</div>
+
 Processing of fragments happens in parallel, while any side effects,
 such as writes into {{GPUBufferBindingType/"storage"|GPUBufferBindingType."storage"}} bindings,
 may happen in any order.
 
+### Depth/Stencil Processing ### {#depth-stencil-processing}
+
+Depth/stencil processing is a fixed-function stage of the render [=pipeline=] that
+outputs the fragment depth and stencil data to be written into the depth/stencil attachment.
+
+<div algorithm data-timeline=queue>
+    <dfn abstract-op>process depth stencil</dfn>(fragment, descriptor, state)
+
+    **Arguments:**
+
+    - |fragment|: The [=Fragment=], produced by [[#fragment-processing]].
+    - |descriptor|: The descriptor of type {{GPURenderPipelineDescriptor}}.
+    - |state|: The active [=RenderState=].
+
+    1. Let |depthStencilDesc| be |descriptor|.{{GPURenderPipelineDescriptor/depthStencil}}.
+
+    1. If |descriptor|.{{GPURenderPipeline/[[writesDepth]]}} is `true` and |fragment|.[=Fragment/depthPassed=] is `true`:
+        1. Set the value of the depth aspect of |state|.{{RenderState/[[depthStencilAttachment]]}} at
+            |fragment|.[=Fragment/destination=] to |fragment|.[=Fragment/depth=].
+
+    1. If |descriptor|.{{GPURenderPipeline/[[writesStencil]]}} is true:
+        1. Set |stencilState| to |depthStencilDesc|.{{GPUDepthStencilState/stencilFront}} if
+            |fragment|.[=Fragment/frontFacing=] is `true` and
+            |depthStencilDesc|.{{GPUDepthStencilState/stencilBack}} otherwise.
+        1. Let |stencilOp| be `undefined`
+        1. If:
+            <dl class=switch>
+                : |fragment|.[=Fragment/stencilPassed=] is `false`:
```

Comment:
This switch statement is confusing as depthPassed is independent of stencilPassed. I think it would be better expressed as if-else chains where we first test stencilFail, depthFail, and then fallback to stencilPass.